src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+
+    public static class Node<E> {
+        private E item;
+        private Node<E> prev;
+        private Node<E> next;
+
+        public Node(Node<E> prev,E item, Node<E> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
+    private Node<T> tail;
+    private Node<T> head;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            Node<T> firstNode = new Node<>(null, value,null);
+            tail = firstNode;
+            head = firstNode;
+            size++;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAddMethod(index);
+        if (index == 0) {
+            if (size == 0) {
+                Node<T> firstNode = new Node<>(null, value,null);
+                tail = firstNode;
+                head = firstNode;
+                size++;
+                return;
+            }
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        }
+        if (index > 0 && index < size) {
+            Node<T> current = searchNodeByIndex(index);
+            Node<T> newNode = new Node<>(current.prev, value, current);
+            current.prev.next = newNode;
+            current.prev = newNode;
+            size++;
+        }
+        if (index == size) {
+            add(value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = searchNodeByIndex(index);
+        return current.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T replacedElement;
+        Node<T> current = searchNodeByIndex(index);
+        replacedElement = current.item;
+        current.item = value;
+        return replacedElement;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = searchNodeByIndex(index);
+        if (size == 1 && index == 0) {
+            size--;
+        } else if (index == 0) {
+            current.next.prev = null;
+            head = current.next;
+            size--;
+        } else if (index > 0 && index < size - 1) {
+            current.prev.next = current.next;
+            current.next.prev = current.prev;
+            size--;
+        } else if (index == size - 1) {
+            current.prev.next = null;
+            tail = current.prev;
+            size--;
+        }
+        return current.item;
```

Comment:
```suggestion
        final T element = current.value;
        Node<T> prev = current.prev;
        Node<T> next = current.next;
        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
            current.prev = null;
        }
        if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
            current.next = null;
        }
        current.value = null;
        size--;
        return element;
```

we could make it simpler